rpcs3/Emu/Cell/Modules/cellCamera.cpp

```diff
@@ -97,109 +111,214 @@ static const char* get_camera_attr_name(s32 value)
 	return nullptr;
 }
 
-// Custom struct to keep track of cameras
-struct camera_t
+static bool check_dev_num(u32 dev_num)
+{
+	return dev_num == 0;
+}
+
+/**
+ * \brief Sets read mode attribute (used for deciding how image data is passed to games)
+ *		  Also sends it to the camera thread
+ *	      NOTE: thread-safe (uses camera_thread::mutex)
+ * \param dev_num Device number (always 0)
+ * \param read_mode Either CELL_CAMERA_READ_FUNCCALL or CELL_CAMERA_READ_DIRECT
+ * \return CELL error code or CELL_OK
+ */
+u32 set_and_send_read_mode(s32 dev_num, const s32 read_mode)
+{
+	if (read_mode == CELL_CAMERA_READ_FUNCCALL ||
+		read_mode == CELL_CAMERA_READ_DIRECT)
+	{
+		if (const auto status = cellCameraSetAttribute(dev_num, CELL_CAMERA_READMODE, read_mode, 0))
+		{
+			return status;
+		}
+	}
+	else
+	{
+		cellCamera.error("Unknown read mode set: %d", read_mode);
+	}
+
+	// Send read mode to camera thread
+	const auto g_camera = fxm::get<camera_thread>();
+
+	g_camera->read_mode.exchange(read_mode);
+
+	return CELL_OK;
+}
+
+std::pair<u32, u32> get_video_resolution(const CellCameraInfoEx& info)
 {
-	struct attr_t
+	std::pair<u32, u32> res;
+	switch (info.resolution)
 	{
-		u32 v1, v2;
-	};
+	case CELL_CAMERA_VGA: return{ 640, 480 };
+	case CELL_CAMERA_QVGA: return { 320, 240 };
+	case CELL_CAMERA_WGA: return{ 640, 360 };
+	case CELL_CAMERA_SPECIFIED_WIDTH_HEIGHT: return{ info.width, info.height };
+	case CELL_CAMERA_RESOLUTION_UNKNOWN:
+	default: return{ 0, 0 };
+	}
+}
+
+u32 get_video_buffer_size(const CellCameraInfoEx& info)
+{
+	u32 width, height;
+	std::tie(width, height) = get_video_resolution(info);
 
-	attr_t attr[500]{};
-};
+	const auto bpp = 4;
+	return width * height * bpp;
+}
+
+/*************************
+ * cellCamera functions  *
+ *************************/
 
 s32 cellCameraInit()
 {
-	cellCamera.warning("cellCameraInit()");
+	cellCamera.todo("cellCameraInit()");
 
 	if (g_cfg.io.camera == camera_handler::null)
 	{
 		return CELL_CAMERA_ERROR_DEVICE_NOT_FOUND;
 	}
 
-	const auto camera = fxm::make<camera_t>();
+	// Start camera thread
+	const auto g_camera = fxm::make<camera_thread>();
 
-	if (!camera)
+	if (!g_camera)
 	{
 		return CELL_CAMERA_ERROR_ALREADY_INIT;
 	}
-
-	switch (g_cfg.io.camera_type)
 	{
```

Comment:
This scope is useless

Reply:
This is a switch statement in old code, I assume you mean my own scope I made for the lock.